### DISCUSSION TITLE
Interlude 1 - the definition of sum1

### DISCUSSION THREAD

**Discussion by Peter-Jeffs** [themetaschemer/malt#27] [Q&A]
Hi, 

I need some help. I'm really enjoying the book and I'm working through the examples using DrRacket. 

I have tried to code sum1 and summed following the outline detailed at the top of page 53 of the book but I'm not getting the results expected. 

Could you please help me by highlighting where I am going wrong? (code as per below).

Thanks, much appreciated.
Peter

(define sum-1
  (lambda (t)
    (summed t (sub1 (tlen t)) 0.0)))

(define summed
  (lambda (t i a)
    (cond
      ((zero? i) (+ (tref t 0) a))
      (else
       (summed t (sub1 i) (+ (tref t i) a))))))

**Comment by dfried00** [comment]
Peter,

Use www.thelittlelearner.com as can be found in the preface.
Then click on Code.
Follow instructions from then on.

**Comment by themetaschemer** [comment]
That code looks correct. What error/difference are you seeing? Could you post your output here? Thanks.

**Reply by Peter-Jeffs** [reply]
Hi, 
Thanks for taking a look at the code. 

Here are my test cases and results:
Test Case 1 (Using Malted 'sum'): Input = (sum (tensor 10.0 12.0 14.0)), result = 36.0
Test Case 2 (Using my version): Input =  (sum-1 (tensor 10.0 12.0 14.0)), result = 36.0
For test cases 1 and 2 my version seems to work as expected. 

However, for test cases 3 and 4, I get different results as follows:
Test Case 3 ( (Using Malted 'sum'): Input = (sum (tensor (tensor (tensor 5.0 8.0) (tensor 12.00 6.00)) (tensor (tensor 7.0 4.0) (tensor 6.0 5.0)))), result = (tensor (tensor 13.0 18.0) (tensor 11.0 11.0))
Test Case 4 ((Using my version): Input =  (sum-1 (tensor (tensor (tensor 5.0 8.0) (tensor 12.00 6.00)) (tensor (tensor 7.0 4.0) (tensor 6.0 5.0)))), result = (tensor (tensor 12.0 12.0) (tensor 18.0 11.0))

Kind regards, Peter

**Reply by Peter-Jeffs** [reply]
and ...
Test Case 7 ( (Using Malted 'sum'): Input = (sum (tensor (tensor 1 2) (tensor 3 4))), result = (tensor 3.0 7.0)
Test Case 4 ((Using my version): Input =(sum-1 (tensor (tensor 1 2) (tensor 3 4))), result = (tensor 4.0 6.0)

Kind Regards, Peter

**Reply by Peter-Jeffs** [reply]
Hi, thank you for your help. I now understand my mistake (as per my response to dfried00). Please ignore the above test cases. Kind Regards, Peter.

**Comment by themetaschemer** [comment]
Also, please make sure you have the latest copy of malt (I don't know when you pulled the repository), and you have your tensor implementation set to learner.

**Reply by Peter-Jeffs** [reply]
Hi, ok. I'll follow your advise. Thank you

**Reply by Peter-Jeffs** [reply]
Hi, 

As suggested, I have applied the following: **racket -e "(require malt/set-impl) (set-impl 'learner)"**

Kind regards, Peter

**Comment by dfried00** [comment]
The sum-1 means its data must be a flat tensor (i.e. like a vector).
When you look at frame 26, where we define sum, you can see the
progression down to a sum-1, and there the tensor is again flat.
Then, on frames 22 and 23, it clarifies that point.

**Reply by Peter-Jeffs** [reply]
Hi, 

Thank you. I think I understand the theory. I was hoping to clarify my understanding by writing some code and validating it against the examples in the book (like I have done with _rank_ and _shape_). I enjoy the 'hands-on' feature of the book it really does help the learning process.

I suppose I'm just curious as to why my code doesn't produce the correct result. 

Kind Regards, Peter

**Comment by dfried00** [comment]
Peter,

Did everything work?

... Dan

**Reply by Peter-Jeffs** [reply]
Hi Dan,

Thank you for your help. I think I now know why I was confused.

I have coded _sum-1_ which when given a flat tensor produces the correct result (as per sum1 in the book), all good so far.

I have **not** yet coded the extended version of sum1 defined as _sum_ which is able to descend into its argument until it finds a tensor one instead of a tensor zero.  My mistake - not sure (now) why I got so confused!? 

Many thanks for your help. Much appreciated.

Kind regards, Peter
